### ISSUE TITLE
VGA text renderer

### ISSUE THREAD

**Issue by Simn** [HaxeFoundation/code-cookbook#55]
anniz on #haxe posted a fun little example of a vga text renderer:

```haxe
class Main {
	static var Font = [
		0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
		0x7E,0x81,0xA5,0x81,0xBD,0x99,0x81,0x7E,
		0x7E,0xFF,0x00,0xFF,0xC3,0xE7,0xFF,0x7E,
		0x6C,0xFE,0xFE,0xFE,0x7C,0x38,0x10,0x00,
		0x10,0x38,0x7C,0xFE,0x7C,0x38,0x10,0x00,
		0x38,0x7C,0x38,0xFE,0xFE,0x92,0x10,0x7C,
		0x00,0x10,0x38,0x7C,0xFE,0x7C,0x38,0x7C,
		0x00,0x00,0x18,0x3C,0x3C,0x18,0x00,0x00,
		0xFF,0xFF,0xE7,0xC3,0xC3,0xE7,0xFF,0xFF,
		0x00,0x3C,0x66,0x42,0x42,0x66,0x3C,0x00,
		0xFF,0xC3,0x99,0xBD,0xBD,0x99,0xC3,0xFF,
		0x0F,0x07,0x0F,0x7D,0xCC,0xCC,0xCC,0x78,
		0x3C,0x66,0x66,0x66,0x3C,0x18,0x7E,0x18,
		0x3F,0x33,0x3F,0x30,0x30,0x70,0xF0,0xE0,
		0x7F,0x63,0x7F,0x63,0x63,0x67,0xE6,0xC0,
		0x99,0x5A,0x3C,0xE7,0xE7,0x3C,0x5A,0x99,
		0x80,0xE0,0xF8,0xFE,0xF8,0xE0,0x80,0x00,
		0x02,0x0E,0x3E,0xFE,0x3E,0x0E,0x02,0x00,
		0x18,0x3C,0x7E,0x18,0x18,0x7E,0x3C,0x18,
		0x66,0x66,0x66,0x66,0x66,0x00,0x66,0x00,
		0x7F,0x00,0x00,0x7B,0x1B,0x1B,0x1B,0x00,
		0x3E,0x63,0x38,0x6C,0x6C,0x38,0x86,0xFC,
		0x00,0x00,0x00,0x00,0x7E,0x7E,0x7E,0x00,
		0x18,0x3C,0x7E,0x18,0x7E,0x3C,0x18,0xFF,
		0x18,0x3C,0x7E,0x18,0x18,0x18,0x18,0x00,
		0x18,0x18,0x18,0x18,0x7E,0x3C,0x18,0x00,
		0x00,0x18,0x0C,0xFE,0x0C,0x18,0x00,0x00,
		0x00,0x30,0x60,0xFE,0x60,0x30,0x00,0x00,
		0x00,0x00,0xC0,0xC0,0xC0,0xFE,0x00,0x00,
		0x00,0x24,0x66,0xFF,0x66,0x24,0x00,0x00,
		0x00,0x18,0x3C,0x7E,0xFF,0xFF,0x00,0x00,
		0x00,0xFF,0xFF,0x7E,0x3C,0x18,0x00,0x00,
		0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
		0x18,0x3C,0x3C,0x18,0x18,0x00,0x18,0x00,
		0x6C,0x6C,0x6C,0x00,0x00,0x00,0x00,0x00,
		0x6C,0x6C,0xFE,0x6C,0xFE,0x6C,0x6C,0x00,
		0x18,0x7E,0xC0,0x7C,0x06,0xFC,0x18,0x00,
		0x00,0xC6,0xCC,0x18,0x30,0x66,0xC6,0x00,
		0x38,0x6C,0x38,0x76,0xDC,0xCC,0x76,0x00,
		0x30,0x30,0x60,0x00,0x00,0x00,0x00,0x00,
		0x18,0x30,0x60,0x60,0x60,0x30,0x18,0x00,
		0x60,0x30,0x18,0x18,0x18,0x30,0x60,0x00,
		0x00,0x66,0x3C,0xFF,0x3C,0x66,0x00,0x00,
		0x00,0x18,0x18,0x7E,0x18,0x18,0x00,0x00,
		0x00,0x00,0x00,0x00,0x00,0x18,0x18,0x30,
		0x00,0x00,0x00,0x7E,0x00,0x00,0x00,0x00,
		0x00,0x00,0x00,0x00,0x00,0x18,0x18,0x00,
		0x06,0x0C,0x18,0x30,0x60,0xC0,0x80,0x00,
		0x7C,0xCE,0xDE,0xF6,0xE6,0xC6,0x7C,0x00,
		0x30,0x70,0x30,0x30,0x30,0x30,0xFC,0x00,
		0x78,0xCC,0x0C,0x38,0x60,0xCC,0xFC,0x00,
		0x78,0xCC,0x0C,0x38,0x0C,0xCC,0x78,0x00,
		0x1C,0x3C,0x6C,0xCC,0xFE,0x0C,0x1E,0x00,
		0xFC,0xC0,0xF8,0x0C,0x0C,0xCC,0x78,0x00,
		0x38,0x60,0xC0,0xF8,0xCC,0xCC,0x78,0x00,
		0xFC,0xCC,0x0C,0x18,0x30,0x30,0x30,0x00,
		0x78,0xCC,0xCC,0x78,0xCC,0xCC,0x78,0x00,
		0x78,0xCC,0xCC,0x7C,0x0C,0x18,0x70,0x00,
		0x00,0x18,0x18,0x00,0x00,0x18,0x18,0x00,
		0x00,0x18,0x18,0x00,0x00,0x18,0x18,0x30,
		0x18,0x30,0x60,0xC0,0x60,0x30,0x18,0x00,
		0x00,0x00,0x7E,0x00,0x7E,0x00,0x00,0x00,
		0x60,0x30,0x18,0x0C,0x18,0x30,0x60,0x00,
		0x3C,0x66,0x0C,0x18,0x18,0x00,0x18,0x00,
		0x7C,0xC6,0xDE,0xDE,0xDC,0xC0,0x7C,0x00,
		0x30,0x78,0xCC,0xCC,0xFC,0xCC,0xCC,0x00,
		0xFC,0x66,0x66,0x7C,0x66,0x66,0xFC,0x00,
		0x3C,0x66,0xC0,0xC0,0xC0,0x66,0x3C,0x00,
		0xF8,0x6C,0x66,0x66,0x66,0x6C,0xF8,0x00,
		0xFE,0x62,0x68,0x78,0x68,0x62,0xFE,0x00,
		0xFE,0x62,0x68,0x78,0x68,0x60,0xF0,0x00,
		0x3C,0x66,0xC0,0xC0,0xCE,0x66,0x3A,0x00,
		0xCC,0xCC,0xCC,0xFC,0xCC,0xCC,0xCC,0x00,
		0x78,0x30,0x30,0x30,0x30,0x30,0x78,0x00,
		0x1E,0x0C,0x0C,0x0C,0xCC,0xCC,0x78,0x00,
		0xE6,0x66,0x6C,0x78,0x6C,0x66,0xE6,0x00,
		0xF0,0x60,0x60,0x60,0x62,0x66,0xFE,0x00,
		0xC6,0xEE,0xFE,0xFE,0xD6,0xC6,0xC6,0x00,
		0xC6,0xE6,0xF6,0xDE,0xCE,0xC6,0xC6,0x00,
		0x38,0x6C,0xC6,0xC6,0xC6,0x6C,0x38,0x00,
		0xFC,0x66,0x66,0x7C,0x60,0x60,0xF0,0x00,
		0x7C,0xC6,0xC6,0xC6,0xD6,0x7C,0x0E,0x00,
		0xFC,0x66,0x66,0x7C,0x6C,0x66,0xE6,0x00,
		0x7C,0xC6,0xE0,0x78,0x0E,0xC6,0x7C,0x00,
		0xFC,0xB4,0x30,0x30,0x30,0x30,0x78,0x00,
		0xCC,0xCC,0xCC,0xCC,0xCC,0xCC,0xFC,0x00,
		0xCC,0xCC,0xCC,0xCC,0xCC,0x78,0x30,0x00,
		0xC6,0xC6,0xC6,0xC6,0xD6,0xFE,0x6C,0x00,
		0xC6,0xC6,0x6C,0x38,0x6C,0xC6,0xC6,0x00,
		0xCC,0xCC,0xCC,0x78,0x30,0x30,0x78,0x00,
		0xFE,0xC6,0x8C,0x18,0x32,0x66,0xFE,0x00,
		0x78,0x60,0x60,0x60,0x60,0x60,0x78,0x00,
		0xC0,0x60,0x30,0x18,0x0C,0x06,0x02,0x00,
		0x78,0x18,0x18,0x18,0x18,0x18,0x78,0x00,
		0x10,0x38,0x6C,0xC6,0x00,0x00,0x00,0x00,
		0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xFF,
		0x30,0x30,0x18,0x00,0x00,0x00,0x00,0x00,
		0x00,0x00,0x78,0x0C,0x7C,0xCC,0x76,0x00,
		0xE0,0x60,0x60,0x7C,0x66,0x66,0xDC,0x00,
		0x00,0x00,0x78,0xCC,0xC0,0xCC,0x78,0x00,
		0x1C,0x0C,0x0C,0x7C,0xCC,0xCC,0x76,0x00,
		0x00,0x00,0x78,0xCC,0xFC,0xC0,0x78,0x00,
		0x38,0x6C,0x64,0xF0,0x60,0x60,0xF0,0x00,
		0x00,0x00,0x76,0xCC,0xCC,0x7C,0x0C,0xF8,
		0xE0,0x60,0x6C,0x76,0x66,0x66,0xE6,0x00,
		0x30,0x00,0x70,0x30,0x30,0x30,0x78,0x00,
		0x0C,0x00,0x1C,0x0C,0x0C,0xCC,0xCC,0x78,
		0xE0,0x60,0x66,0x6C,0x78,0x6C,0xE6,0x00,
		0x70,0x30,0x30,0x30,0x30,0x30,0x78,0x00,
		0x00,0x00,0xCC,0xFE,0xFE,0xD6,0xD6,0x00,
		0x00,0x00,0xB8,0xCC,0xCC,0xCC,0xCC,0x00,
		0x00,0x00,0x78,0xCC,0xCC,0xCC,0x78,0x00,
		0x00,0x00,0xDC,0x66,0x66,0x7C,0x60,0xF0,
		0x00,0x00,0x76,0xCC,0xCC,0x7C,0x0C,0x1E,
		0x00,0x00,0xDC,0x76,0x62,0x60,0xF0,0x00,
		0x00,0x00,0x7C,0xC0,0x70,0x1C,0xF8,0x00,
		0x10,0x30,0xFC,0x30,0x30,0x34,0x18,0x00,
		0x00,0x00,0xCC,0xCC,0xCC,0xCC,0x76,0x00,
		0x00,0x00,0xCC,0xCC,0xCC,0x78,0x30,0x00,
		0x00,0x00,0xC6,0xC6,0xD6,0xFE,0x6C,0x00,
		0x00,0x00,0xC6,0x6C,0x38,0x6C,0xC6,0x00,
		0x00,0x00,0xCC,0xCC,0xCC,0x7C,0x0C,0xF8,
		0x00,0x00,0xFC,0x98,0x30,0x64,0xFC,0x00,
		0x1C,0x30,0x30,0xE0,0x30,0x30,0x1C,0x00,
		0x18,0x18,0x18,0x00,0x18,0x18,0x18,0x00,
		0xE0,0x30,0x30,0x1C,0x30,0x30,0xE0,0x00,
		0x76,0xDC,0x00,0x00,0x00,0x00,0x00,0x00,
		0x00,0x10,0x38,0x6C,0xC6,0xC6,0xFE,0x00
	];

	public static function PcRenderLetter(charCode:Int, colorOn:Int, colorOff:Int, baseRow:Int, baseCol:Int, targetBMD:flash.display.BitmapData) {
		var baseIndex = charCode << 3;
		for (curRow in 0...8) {
			var rowBits = Font[baseIndex + curRow];
			for (curCol in 0...8) {
				targetBMD.setPixel32(baseCol+curCol, baseRow+curRow, (((rowBits << curCol) & 0x80) == 0x80)? colorOn : colorOff);
			}
		}
	}

	public static function main() {
		var bmd = new flash.display.BitmapData(320, 200, false, 0);
		var bm = new flash.display.Bitmap(bmd);

		var msgLines = [
			"256KB OK",
			"",
			"",
			"The current date is: Thu 04/20/1989",
			"Enter new date: (mm-dd-yy)",
			"",
			"The current time is: 03:14:15.9",
			"Enter new time:",
			"",
			"",
			"MS-DOS 2.04",
			"",
			"A:\\>"
		];

		for (j in 0...msgLines.length) {
			var msgStr = msgLines[j];
			for (i in 0...msgStr.length) {
				var c = msgStr.charCodeAt(i);
				PcRenderLetter(c, 0xa0a0a0, 0, (j>>1) + (j<<3), (i>>1) + (i<<3), bmd);
			}
		}
		flash.Lib.current.addChild(bm);
	}
}
```

Maybe we could port it to canvas?

**Comment by markknol** [comment]
http://try.haxe.org/#1EDC7

**Comment by fullofcaffeine** [comment]
Could someone elaborate on how he went about creating/designing that font with an array of hex numbers? I know it has to do with a bitmap, but I can't really visualize it. How would someone design a font out of the blue with only an array of bytes? That's mindblowing to me :)

**Comment by fullofcaffeine** [comment]
Is that array of bytes a bitmap image with a grid of chars?

**Comment by dstrekelj** [comment]
A character is 8x8 pixels in size. It can therefore be represented by eight octets, where each octet is a row (or column, depending on design) of the monochrome 8x8 character bitmap. The replicated font in this example is the IBM BIOS VGA font (or a variant of it). You can read more about it [here][1].

The character bitmaps are stored in the `Font` array ordered by their character codes. It can be said that the character code encodes in itself the base row and base column of the character bitmap. Because the bitmaps are stored in groups of eight due to the character size, we can observe that every eighth entry - if counting from 0 - represents the bitmap of a different character.

Thus, when a text string is rendered, it is probed for its character code. The character code is decoded into a `Font` array index through a left bit shift by three bits, due to the characters being packaged by eight octets. From then on, every row and column of the character bitmap grid is traversed, and the colour value sampled, before the image's pixel colour at the given position is set. For a more detailed explanation on how to display a character, see [here][2].

The position at which the character is drawn is determined from the character's position in the text string, and the size of the character bitmap. This can also be achieved through bit shifting or a modulus operation.

All in all, it's quite a cool little example.

[1]: http://int10h.org/oldschool-pc-fonts/fontlist/
[2]: http://wiki.osdev.org/VGA_Fonts

**Comment by fullofcaffeine** [comment]
Wow, thanks for the explanation! That's quite a bit of "black magic", but I'm starting to understand :)

>Thus, when a text string is rendered, it is probed for its character code. The character code is decoded into a Font array index through a left bit shift by three bits due to the characters being packaged by eight octets

Why so? Why three bits in this case?

**Comment by dstrekelj** [comment]
@fullofcaffeine You're welcome!

It all boils down to how base-10 (decimal) numbers are represented in base-2 (binary). I trust you know the basics so I'll skip the fundamentals. If I went down the numeral system rabbit hole, I don't think I'd be able to find my way out :)

Much like decimal numbers have their own set of operators and operations - such as addition, multiplication, division, and subtraction - binary numbers define their own bitwise operations. One category of such operations are _shifting operations_.

Shifting operations "shift" a binary number to either the left or right by a specified number of bits (positions, or numerals in the binary number). Due to the nature of signed binary numbers, shifting can also be signed or unsigned - but that is beside the point.

Because every bit of a binary number can be observed as a power-of-two value (2^n, where n is the bit position in number in range [0, number of bits - 1]), a shift to the left or right has the same effect as a multiplication or division (respectively) by 2^n, where n is the number of places shifted. The result of shifting depends on convention - whether the most / least significant bit (highest power-of-two value in relation to position in number) is placed first / last in the binary number.

If, for example, we consider the leftmost bit to be the most significant, then shifting to the left by 3 positions is equal to multiplying the decimal representation of the number by 2^3 (or 8). Shifting to the right by 3 positions, however, results in a division by 2^3.

In short, shifting operations are a way of performing integer multiplication or division by power-of-two values.

These days any decent compiler should optimise multiplications and divisions by power-of-two values into a bit shift operation. That is because shifting, being a logical operation, is several cycles faster than an arithmetic operation. But that is a different rabbit hole, for another time :)

**Comment by dstrekelj** [comment]
@markknol Would you mind if I created a cookbook entry out of this? Of course, you would be credited with porting the original code.

**Comment by markknol** [comment]
Yes do it!! That would be a nice addition!

**Comment by dstrekelj** [comment]
I guess this issue can be closed now :)

**Comment by fullofcaffeine** [comment]
@dstrekelj Once again, thank you very much for the comprehensive explanation!

**Comment by fullofcaffeine** [comment]
I'd be pretty cool if we had more of that retro/oldschool/lowlevel kind of examples, other stuff that would be very educational (gamedev-wise):

* A basic emulator;
* synthetizing 8/16 bit souds;
* Simulating how 8/16 bit consoles rendered sprites (related to emulator above, maybe);

Thanks for this contribution! The page on the cookbook has been expanded and is really informative!
